### ISSUE TITLE
Xcode 7.1-beta2: Signal observe crash with multiple observers

### ISSUE THREAD

**Issue by jsslai** [ReactiveCocoa/ReactiveCocoa#2417]
Following code crashes after some iterations: 

```
let signal: Signal<Void, NoError> = Signal { _ in
    return nil
}

while(true) {
   let disposable = signal.observeCompleted {}
   disposable?.dispose()
}
```

Console log:
```
Test(78365,0x107691300) malloc: *** mach_vm_map(size=70918499991552) failed (error code=3)
*** error: can't allocate region
*** set a breakpoint in malloc_error_break to debug
```

It crashes on this line: https://github.com/ReactiveCocoa/ReactiveCocoa/blob/v4.0-alpha.1/ReactiveCocoa/Swift/Bag.swift#L35

Is this a Swift or ReactiveCocoa bug or am I doing something wrong?

**Comment by neilpa** [comment]
Seems like a Swift/ARC bug. How many iterations before it actually dies though?

**Comment by neilpa** [comment]
This must be a boatload of iterations, you're trying to allocate a monstrous amount of memory - `mach_vm_map(size=70918499991552)`

**Comment by NachoSoto** [comment]
Yeah, this is way more than just *some* iterations. Because it's an infinite loop there is no chance for the autorelease pool to be drained, so you're just allocating more and more memory.

What are you trying to do exactly?

**Comment by jsslai** [comment]
It was maybe 40 iterations. I first noticed this with my app and the crash happened after I repeatedly opened a view controller which is observing a signal 8 times.

**Comment by jsslai** [comment]
That is only a small example code how to repeat the crash. Like I said, this first happened with my view controller. I'm observing a signal in viewDidAppear and calling dispose in viewDidDisappear.

**Comment by NachoSoto** [comment]
That's funny, I just saw this exact same crash in my app, and I'm not doing anything too big... I'd say I don't have more than 100 observers total on several signals.

**Comment by NachoSoto** [comment]
<img width="1101" alt="screen shot 2015-09-26 at 20 17 16" src="https://cloud.githubusercontent.com/assets/685609/10120983/a4a2fa1e-648b-11e5-9885-50e1fcb8060f.png">

It's not because I'm allocating too much memory though!

**Comment by NachoSoto** [comment]
I wonder if @dapenggao saw this at all in #2386 while running the benchmarks?

**Comment by dapenggao** [comment]
@NachoSoto Interesting! This never happened when running the benchmarks.

**Comment by NachoSoto** [comment]
I'm running it with the address sanitizer and it's picking up something, but having trouble to actually see it because Xcode crashes right after...

**Comment by NachoSoto** [comment]
Managed to take a screenshot before crashing:
<img width="1792" alt="screen shot 2015-09-26 at 21 11 47" src="https://cloud.githubusercontent.com/assets/685609/10121151/3ea2c048-6493-11e5-87fb-0045f7869463.png">

It doesn't tell us anything new though...

**Comment by NachoSoto** [comment]
There's something interesting in that stack: notice the `_forceCreateUniqueMutableBuffer`. Every time we modify the bag we're copying the entire array `elements`, over and over again... I'm not sure why the memory in Xcode doesn't indicate otherwise, but I wouldn't be surprised if we're allocating too much memory in the heap due to this.

Maybe we can start by making `Bag` a class (to avoid extra copies). We also use it inside of an `Atomic<Bag>` anyway!

**Comment by dapenggao** [comment]
@NachoSoto I thought that Swift performs optimizations to avoid actually copying the entire array if the it is uniquely referenced. Isn't this the case here?

**Comment by NachoSoto** [comment]
@dapenggao no, because of our use of `Atomic` :/

**Comment by NachoSoto** [comment]
I added a similar test to @jsslai's in `SignalSpec.swift`:
```swift
let (signal, sink) = Signal<Int, NoError>.pipe()

for i in 0..<50 {
	print(i)

	autoreleasepool {
		let disposable = signal.observe { _ in }

		sendNext(sink, 10)

		disposable!.dispose()
	}
}
```

This crashes after 39 iterations, currently digging into it...

**Comment by NachoSoto** [comment]
I just ran the same test with `Xcode 7.0` (so `Swift 2.0` instead of `Swift 2.1`) and it passes!

**Comment by NachoSoto** [comment]
Filed: http://www.openradar.me/22873620

**Comment by mdiep** [comment]
#2421 added test coverage for this and noted that the bug was fixed in Xcode 7.1b3.
